src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -142,14 +181,25 @@ public synchronized void add(DecoratedKey key,
 
                 try
                 {
-                    data.putSingleton(encodedTerm, primaryKey, primaryKeysReducer, term.limit() <= MAX_RECURSIVE_KEY_LENGTH);
+                    data.putSingleton(encodedTerm, primaryKey, (existing, update) -> {
+                        // First do the normal primary keys reduction
+                        PrimaryKeys result = primaryKeysReducer.apply(existing, update);
+
+                        // Then update term frequency
+                        var pkbc = new PkWithTerm(update, encodedTerm);
+                        termFrequencies.merge(pkbc, 1, Integer::sum);
+
+                        return result;
+                    }, term.limit() <= MAX_RECURSIVE_KEY_LENGTH);
                 }
                 catch (TrieSpaceExhaustedException e)
                 {
                     Throwables.throwAsUncheckedException(e);
                 }
             }
 
+            docLengths.put(primaryKey, tokenCount);
```

Comment:
It seems worth skipping this for non-analyzed columns.

This class generally attempts to keep track of memory utilization to then influence when we need to flush, IIUC. It seems like we should know update the `onHeapAllocationsTracker` with the relevant number of bytes when we add an entry.